frontend/ui/elements/common_settings_menu_table.lua

```diff
@@ -579,6 +600,37 @@ common_settings.document_metadata_location = {
         },
         genMetadataFolderMenuItem("doc"),
         genMetadataFolderMenuItem("dir"),
+        genMetadataFolderMenuItem("hash"),
+        { -- hash-based metadata count / TextViewer
+            text_func = function()
+                local hash_text = _("Show documents with hash-based metadata")
+                local no_hash_text = _("No documents with hash-based metadata")
+                if G_reader_settings:readSetting("document_metadata_folder") == "hash" then
+                    if DocSettings.isHashLocationEnabled() then return hash_text end
+                    -- displays if the hash location setting is selected but no hash-based metadata is found
+                    return no_hash_text
+                end
+                if DocSettings.isHashLocationEnabled() then return  "⚠ "..hash_text end
+                return no_hash_text
+            end,
+            keep_menu_open = true,
+            enabled_func = function()
+                return DocSettings.isHashLocationEnabled()
+            end,
+            callback = function()
+                local hash_file_infos = DocSettings.getHashDirSdrInfos()
+                local book_info_list = ""
+                for i, file_info in ipairs(hash_file_infos) do
+                    book_info_list = book_info_list.."\n"..i..". "..file_info
```

Comment:
Yes, in a loop table.concat is the preferred pattern. Though I don't know if this particular loop is big enough for it to matter?

Reply:
PS I always keep forgetting about https://luajit.org/ext_buffer.html

> The string buffer library allows high-performance manipulation of string-like data.
> 
> Unlike Lua strings, which are constants, string buffers are mutable sequences of 8-bit (binary-transparent) characters. Data can be stored, formatted and encoded into a string buffer and later converted, extracted or decoded.
> 
> The convenient string buffer API simplifies common string manipulation tasks, that would otherwise require creating many intermediate strings. String buffers improve performance by eliminating redundant memory copies, object creation, string interning and garbage collection overhead. In conjunction with the FFI library, they allow zero-copy operations.
> 
> The string buffer library also includes a high-performance [serializer](https://luajit.org/serialize) for Lua objects.

Reply:
Added